src/snowflake/cli/api/project/schemas/project_definition.py

```diff
@@ -203,22 +179,79 @@ def apply_mixins(cls, data: Dict) -> Dict:
         if "mixins" not in data or "entities" not in data:
             return data
 
-        for entity in data["entities"].values():
+        entities = data["entities"]
+        for entity_name, entity in entities.items():
             entity_mixins = entity_mixins_to_list(
                 entity.get("meta", {}).get("use_mixins")
             )
 
-            entity_fields = get_allowed_fields_for_entity(entity)
-            if entity_fields and entity_mixins:
-                for mixin_name in entity_mixins:
-                    if mixin_name in data["mixins"]:
-                        for key, value in data["mixins"][mixin_name].items():
-                            if key in entity_fields:
-                                entity[key] = value
-                    else:
-                        raise ValueError(f"Mixin {mixin_name} not found in mixins")
+            merged_values = cls._merge_mixins_with_entity(
+                entity_name=entity_name,
+                entity=entity,
+                entity_mixins_names=entity_mixins,
+                mixins=data["mixins"],
+            )
+            entities[entity_name] = merged_values
+        return data
+
+    @staticmethod
+    def _merge_mixins_with_entity(
+        entity_name: str, entity: dict, entity_mixins_names: list, mixins: dict
+    ) -> dict:
+        # Validate mixins
+        for mixin_name in entity_mixins_names:
+            if mixin_name not in mixins:
+                raise ValueError(f"Mixin {mixin_name} not defined")
+
+        # Build object override data from mixins
+        data: dict = {}
+        for mx_name in entity_mixins_names:
+            data = DefinitionV20._merge_data(data, mixins[mx_name])
+
+        for key, override_value in data.items():
+            if key not in get_allowed_fields_for_entity(entity):
+                raise ValueError(
+                    f"Unsupported key '{key}' for entity of type {entity['type']} "
+                )
+
+            entity_value = entity.get(key)
+            if entity_value is not None and not isinstance(
+                entity_value, type(override_value)
+            ):
+                raise ValueError(
+                    f"Value from mixins for property {key} is of type '{type(override_value).__name__}' "
+                    f"while entity {entity_name} expects value of type '{type(entity_value).__name__}'"
+                )
+
+        # Apply entity data on top of mixins
+        data = DefinitionV20._merge_data(data, entity)
         return data
 
+    @staticmethod
+    def _merge_data(
+        left: dict | list | scalar | None,
+        right: dict | list | scalar | None,
+    ):
+        if left is None:
+            return right
+
+        # At that point left and right are of the same type
+        if isinstance(left, dict) and isinstance(right, dict):
+            data = dict(left)
+            for key in right:
+                data[key] = DefinitionV20._merge_data(
```

Comment:
Non-blocking nit: as a matter of code style I think `@classmethod` and `cls` would be more appropriate to avoid having to reference the name of the class multiple times

Reply:
Good point, fixed

Reply:
This one `DefinitionV20` got missed!